src/problems/AscotProblem.C

```diff
@@ -51,7 +52,117 @@ AscotProblem::syncSolutions(Direction direction)
+    H5::Attribute results_active_attr = results_group.openAttribute("active");
+    // Get its string type
+    StrType stype = results_active_attr.getStrType();
+    // Read the active run number into a string buffer
+    std::string active_result_num;
+    results_active_attr.read(stype, active_result_num);
+    // Open the active run group
+    std::string endstate_name = "run_" + active_result_num + "/endstate";
+    Group endstate_group = results_group.openGroup(endstate_name);
+    return endstate_group;
+  }
+  else
+  {
+    throw MooseException("ASCOT5 HDF5 File missing 'active' attribute.");
+  }
+}
+
+std::vector<int64_t>
+AscotProblem::getWallTileHits(Group & endstate_group)
+{
+  // Open the walltile dataset
+  DataSet walltile_dataset = endstate_group.openDataSet("walltile");
+  // Get the walltile dataset's data space
+  DataSpace walltile_dataspace = walltile_dataset.getSpace();
+  // check we only have 1 dim
+  if (walltile_dataspace.getSimpleExtentNdims() == 1)
+  {
+    hssize_t n_markers = walltile_dataspace.getSimpleExtentNpoints();
+    std::vector<int64_t> walltile(n_markers);
+    walltile_dataset.read(walltile.data(), PredType::NATIVE_INT64);
+    return walltile;
+  }
+  else
+  {
+    throw MooseException("ASCOT5 HDF5 File walltile dataset of incorrect dim.");
+  }
+}
+
+std::vector<double_t>
+AscotProblem::getParticleEnergies(Group & endstate_group)
+{
+
+  std::vector<double_t> particle_energies;
+  // Open the one of the datasets to check the number of ions/markers
+  DataSet mass_dataset = endstate_group.openDataSet("mass");
+  DataSpace mass_dataspace = mass_dataset.getSpace();
+  // Check we only have 1 dim
+  if (mass_dataspace.getSimpleExtentNdims() != 1)
+  {
+    throw MooseException("ASCOT5 HDF5 File walltile dataset of incorrect dim.");
+  }
+  const hsize_t n_markers = mass_dataspace.getSimpleExtentNpoints();
+
+  // Read in mass
+  std::vector<double_t> mass(n_markers);
+  mass_dataset.read(mass.data(), PredType::NATIVE_DOUBLE);
+  // Read in velocities
+  std::unordered_map<std::string, std::vector<double_t>> velocities;
+  std::vector<std::string> velocity_names{"vr", "vphi", "vz"};
+  for (auto && name : velocity_names)
+  {
+    velocities[name].resize(n_markers);
+    DataSet v_dataset = endstate_group.openDataSet(name);
+    v_dataset.read(velocities[name].data(), PredType::NATIVE_DOUBLE);
+  }
+
+  std::vector<double_t> velocity;
+  for (hsize_t i = 0; i < n_markers; i++)
+  {
+    for (auto && name : velocity_names)
+    {
+      velocity.push_back(velocities[name][i]);
+    }
+    particle_energies.push_back(calculateRelativisticEnergy(mass[i], velocity.data()));
+    velocity.clear();
+  }
+
+  return particle_energies;
+}
+
+double_t
+AscotProblem::calculateRelativisticEnergy(double_t mass, double_t velocity[3])
+{
+  double_t magnitude = std::inner_product(velocity, velocity + 3, velocity, 0.0);
+  double_t gamma = 1.0 / sqrt(1.0 - magnitude / pow(constants::c, 2.0));
```

Comment:
Point taken about numerical accuracy, and I agree that is probably the safer route (||v/c|| rather than ||v|| / c^2). Let me know what you think of my updated implementation with the `transform()` and lambda expression.

Thanks for the info about the `pow()` function. Helpful for below and in the future.